java-backend/src/main/java/org/kframework/backend/java/symbolic/ConjunctiveFormula.java

```diff
@@ -386,127 +386,130 @@ private ConjunctiveFormula simplify(boolean patternFolding, boolean partialSimpl
-                                eliminationSubstitution,
-                                context);
-                        change = true;
-                        if (substitution.isFalse(global)) {
+                            substitution = ImmutableMapSubstitution.composeAndEvaluate(
+                                    substitution,
+                                    eliminationSubstitution,
+                                    context);
+                            change = true;
+                            if (substitution.isFalse(global)) {
+                                return falsify(substitution, equalities, disjunctions, equality);
+                            }
+                        } else if (leftHandSide instanceof Variable
+                                && rightHandSide.isNormal()
+                                && rightHandSide.variableSet().contains(leftHandSide)) {
+                            // occurs
                             return falsify(substitution, equalities, disjunctions, equality);
+                        } else if (rightHandSide instanceof Variable
+                                && leftHandSide.isNormal()
+                                && leftHandSide.variableSet().contains(rightHandSide)) {
+                            // swap + occurs
+                            return falsify(substitution, equalities, disjunctions, equality);
+                        } else {
+                            // unsimplified equation
+                            pendingEqualities = pendingEqualities.plus(equality);
                         }
-                    } else if (leftHandSide instanceof Variable
-                            && rightHandSide.isNormal()
-                            && rightHandSide.variableSet().contains(leftHandSide)) {
-                        // occurs
-                        return falsify(substitution, equalities, disjunctions, equality);
-                    } else if (rightHandSide instanceof Variable
-                            && leftHandSide.isNormal()
-                            && leftHandSide.variableSet().contains(rightHandSide)) {
-                        // swap + occurs
-                        return falsify(substitution, equalities, disjunctions, equality);
-                    } else {
-                        // unsimplified equation
-                        pendingEqualities = pendingEqualities.plus(equality);
                     }
                 }
-            }
-            equalities = pendingEqualities;
-        } while (change);
+                equalities = pendingEqualities;
+            } while (change);
 
-        context.setTopConstraint(originalTopConstraint);
-        return ConjunctiveFormula.of(substitution, equalities, disjunctions, global);
+            return ConjunctiveFormula.of(substitution, equalities, disjunctions, global);
+        } finally {
+            context.setTopConstraint(originalTopConstraint);
+        }
```

Comment:
Github diff is not smart enough to show the meaningful difference here. :(

So, the main difference is the addition of the `finally` block here? Or, do you have other changes inside the `try` block?

Reply:
If the `finally` block is the only change made, then it looks good to me, and you can merge.

Reply:
Yes, finally block is the only change.